src/pool/pool_disjoint.c

```diff
@@ -0,0 +1,1171 @@
+    // NOTE: this function must be called under bucket->bucket_lock
+    slab_list_item_t *slab_it = slab->iter;
+    assert(slab_it->val != NULL);
+    if (bucket_can_pool(bucket, to_pool)) {
+        DL_DELETE(bucket->unavailable_slabs, slab_it);
+        slab_it->prev = NULL;
+        DL_PREPEND(bucket->available_slabs, slab_it);
+        bucket->available_slabs_num++;
+    } else {
+        slab_unreg(slab_it->val);
+        DL_DELETE(bucket->unavailable_slabs, slab_it);
+        destroy_slab(slab_it->val);
+    }
+}
+
+slab_list_item_t *bucket_get_avail_slab(bucket_t *bucket, bool *from_pool) {
+    if (bucket->available_slabs == NULL) {
+        slab_t *slab = create_slab(bucket, false /* chunked */);
+        if (slab == NULL) {
+            LOG_ERR("create_slab failed!")
+            return NULL;
+        }
+
+        slab_reg(slab);
+        DL_PREPEND(bucket->available_slabs, slab->iter);
+        bucket->available_slabs_num++;
+        bucket_update_stats(bucket, 1, 0);
+        *from_pool = false;
+    } else {
+        slab_t *slab = bucket->available_slabs->val;
+        if (slab->num_allocated == 0) {
+            // If this was an empty slab, it was in the pool.
+            // Now it is no longer in the pool, so update count.
+            --bucket->chunked_slabs_in_pool;
+            bucket_decrement_pool(bucket, from_pool);
+        } else {
+            // Allocation from existing slab is treated as from pool for statistics.
+            *from_pool = true;
+        }
+    }
+
+    return bucket->available_slabs;
+}
+
+size_t bucket_capacity(bucket_t *bucket) {
+    // For buckets used in chunked mode, just one slab in pool is sufficient.
+    // For larger buckets, the capacity could be more and is adjustable.
+    if (bucket->size <= bucket_chunk_cut_off(bucket)) {
+        return 1;
+    } else {
+        return bucket->pool->params.capacity;
+    }
+}
+
+void bucket_update_stats(bucket_t *bucket, int in_use, int in_pool) {
+    if (bucket->pool->params.pool_trace == 0) {
+        return;
+    }
+
+    bucket->curr_slabs_in_use += in_use;
+    bucket->max_slabs_in_use =
+        utils_max(bucket->curr_slabs_in_use, bucket->max_slabs_in_use);
+
+    bucket->curr_slabs_in_pool += in_pool;
+    bucket->max_slabs_in_pool =
+        utils_max(bucket->curr_slabs_in_pool, bucket->max_slabs_in_pool);
+
+    // Increment or decrement current pool sizes based on whether
+    // slab was added to or removed from pool.
+    bucket->pool->params.cur_pool_size +=
+        in_pool * bucket_slab_alloc_size(bucket);
+}
+
+void bucket_decrement_pool(bucket_t *bucket, bool *from_pool) {
+    // If a slab was available in the pool then note that the current pooled
+    // size has reduced by the size of a slab in this bucket.
+    *from_pool = true;
+    bucket_update_stats(bucket, 1, -1);
+    utils_fetch_and_add64(&bucket->shared_limits->total_size,
+                          -(long long)bucket_slab_alloc_size(bucket));
+}
```

Comment:
I do not understand why we are decreasing shared limts here, and how it is supposed to work. My guess it is not working at all. Shared limits according to documentation are supposed to "secifies hard limit for memory allocated from a provider." why we are using it like limit of free memory not returned to provider? IDK